src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,387 @@
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+    // We received arguments from the client, deserialize them back to JS.
+    if (serializedArgs.size() > 0) {
+      auto args = KJ_REQUIRE_NONNULL(
+          deserializeV8(js, kj::heapArray(serializedArgs)).tryCast<jsg::JsArray>(),
+          "expected JsArray when deserializing arguments.");
+      // Call() expects a `Local<Value> []`... so we populate an array.
+      auto arguments = kj::heapArray<v8::Local<v8::Value>>(args.size());
+      for (size_t i = 0; i < args.size(); ++i) {
+        arguments[i] = args.get(js, i);
+      }
+      return jsg::check(fn->Call(js.v8Context(), fn, args.size(), arguments.begin()));
+    } else {
+      return jsg::check(fn->Call(js.v8Context(), fn, 0, nullptr));
+    }
+  };
+
+  // The serialized JS value and a bool to discern whether we've thrown an exception.
+  // Serializing may throw if the data we are serializing is not structured-cloneable.
+  // In this case, we will just serialize the exception message and set `caughtException`.
+  struct SerializedResult {
+    kj::Array<kj::byte> data;
+    bool caughtException;
+  };
+
+  // Serializes the return value of our method.
+  SerializedResult serialize(jsg::Lock& js, jsg::JsValue value) {
+    kj::Maybe<SerializedResult> result;
+
+    js.tryCatch([&] {
+      auto data = serializeV8(js, value.addRef(js).getHandle(js));
+      // Serialized okay!
+      result = SerializedResult { .data = kj::mv(data), .caughtException = false };
+    }, [this, &js, &result](jsg::Value exception) {
+      auto fullError = kj::str(jsg::check(exception.getHandle(js)->ToString(js.v8Context())));
+
+      // We may need to slice part of the message so we don't have `Error: Error: msg`.
+      kj::Maybe<kj::StringPtr> errorMessage(fullError);
+      if (fullError.startsWith("Error: "_kj)) {
+        errorMessage = fullError.slice(kj::size("Error: "_kj));
+      }
+
+      // Failed to serialize data.
+      auto failedResult = serialize(js, js.error(KJ_REQUIRE_NONNULL(errorMessage)));
+      // We are in the error branch, so we must set caughtException.
+      failedResult.caughtException = true;
+      result = kj::mv(failedResult);
+    });
+    return kj::mv(KJ_REQUIRE_NONNULL(result));
+  }
+
+  // Invokes the given function and serializes the result for RPC.
+  kj::Promise<SerializedResult> callFuncAndSerialize(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+
+    kj::Maybe<v8::Local<v8::Value>> eventualResult;
+    bool caughtException = false;
+    js.tryCatch([&] {
+      eventualResult = invokeFn(js, fn, serializedArgs);
+    }, [&](jsg::Value exception) {
+      // We threw while executing JS.
+      caughtException = true;
+      eventualResult = exception.getHandle(js);
+    });
+
+    auto result = KJ_REQUIRE_NONNULL(eventualResult);
+    if (result->IsPromise()) {
```

Comment:
If you're able to wrap the function in a `jsg::Function<T>` where `T` is a function signature that returns a `jsg::Promise<jsg::Value>`, then the function call will always return a promise even if the function returns synchronously. Makes it a lot easier to skip this check.

Reply:
So, I didn't do _exactly_ this, since the `jsg::Function` stuff got a bit messy, but I'm wrapping the result of the JS func call in a promise now and skipping this step ([starting here](https://github.com/cloudflare/workerd/compare/b249885946d71f10959e36108c3ee43d10858753..f12f95f6a9e2c4ce0fd51d7c21e54c46276ae140#diff-bbe685e15c44d4de0aed0b4b484f1c0fbe1427fccbf80ab9f65c6180f6632203R316)).